### PR TITLE
Update clang-tidy checks for recent clang versions

### PR DESCRIPTION
**Pull request by masterleinad**
Related to #14589.

Both, `google-readability-casting` and `modernize-return-braced-init-list` produce a lot of warnings with little (controversial) benefit. `misc-size of-container` doesn't exist anymore but using `misc-definitions-in-headers` as a dummy check seems to work.

### CODE REVIEW

File: source/grid/grid_generator.cc [hunk 1]
```diff
@@ -5505,14 +5505,14 @@ namespace GridGenerator
...
-        vertices.emplace_back(Point<3>(-d, height, d));
-        vertices.emplace_back(Point<3>(d, height, d));
+        vertices.emplace_back(-d, height, -d);
+        vertices.emplace_back(d, height, -d);
+        vertices.emplace_back(-a, height, -a);
+        vertices.emplace_back(a, height, -a);
+        vertices.emplace_back(-a, height, a);
+        vertices.emplace_back(a, height, a);
+        vertices.emplace_back(-d, height, d);
+        vertices.emplace_back(d, height, d);
```
Comment (drwells) [inline source/grid/grid_generator.cc:5515]: Much nicer!

File: source/lac/sparsity_tools.cc [hunk 2]
```diff
@@ -621,9 +621,8 @@ namespace SparsityTools
     DynamicSparsityPattern::size_type next_free_number = 0;
 
     // enumerate the first round dofs
-    for (DynamicSparsityPattern::size_type i = 0; i != last_round_dofs.size();
-         ++i)
-      new_indices[last_round_dofs[i]] = next_free_number++;
+    for (unsigned int last_round_dof : last_round_dofs)
```
Comment (drwells) [inline source/lac/sparsity_tools.cc:624]:
```suggestion
    for (const auto &last_round_dof : last_round_dofs)
```

**Review by drwells — APPROVED** [review]
Looks good - this should make the CI happy.

**Comment by masterleinad** [comment]
For some reason, `clang-tidy` complains about the casting between unrelated types for `MPI_STATUS[ES]_IGNORE` in the CI. Even casting explicitly doesn't help. The last commit removes that check (which was introduced in the first commit).

**Pull request merged**